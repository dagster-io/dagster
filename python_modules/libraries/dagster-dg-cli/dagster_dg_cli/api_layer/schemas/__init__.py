"""Pydantic schemas for Dagster Plus API."""
