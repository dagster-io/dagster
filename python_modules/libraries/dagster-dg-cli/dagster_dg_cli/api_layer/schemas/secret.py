"""Secret models for REST-like API."""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class DgApiSecretScope(str, Enum):
    """Secret scope enum for filtering."""

    DEPLOYMENT = "deployment"
    ORGANIZATION = "organization"


class DgApiUpdatedBy(BaseModel):
    """User who last updated the secret."""

    email: str


class DgApiSecret(BaseModel):
    """Secret resource model."""

    id: str
    name: str = Field(alias="secretName")  # GraphQL field is secretName
    value: Optional[str] = Field(
        alias="secretValue", default=None
    )  # Hidden by default for security
    location_names: list[str] = Field(alias="locationNames")
    full_deployment_scope: bool = Field(alias="fullDeploymentScope")
    all_branch_deployments_scope: bool = Field(alias="allBranchDeploymentsScope")
    specific_branch_deployment_scope: Optional[str] = Field(
        alias="specificBranchDeploymentScope", default=None
    )
    local_deployment_scope: bool = Field(alias="localDeploymentScope")
    can_view_secret_value: bool = Field(alias="canViewSecretValue")
    can_edit_secret: bool = Field(alias="canEditSecret")
    updated_by: Optional[DgApiUpdatedBy] = Field(alias="updatedBy", default=None)
    update_timestamp: Optional[datetime] = Field(alias="updateTimestamp", default=None)

    class Config:
        from_attributes = True  # For future ORM compatibility
        populate_by_name = True  # Allow both alias and field name (Pydantic V2)


class DgApiSecretList(BaseModel):
    """GET /api/secrets response."""

    items: list[DgApiSecret]
    total: int


class DgApiSecretScopesInput(BaseModel):
    """GraphQL SecretScopesInput type for filtering."""

    full_deployment_scope: Optional[bool] = Field(alias="fullDeploymentScope", default=None)
    all_branch_deployments_scope: Optional[bool] = Field(
        alias="allBranchDeploymentsScope", default=None
    )
    specific_branch_deployment_scope: Optional[str] = Field(
        alias="specificBranchDeploymentScope", default=None
    )
    local_deployment_scope: Optional[bool] = Field(alias="localDeploymentScope", default=None)

    class Config:
        populate_by_name = True  # Pydantic V2 equivalent of allow_population_by_field_name

    def to_dict(self) -> dict:
        """Convert to dict for GraphQL variables, excluding None values."""
        result = {}
        if self.full_deployment_scope is not None:
            result["fullDeploymentScope"] = self.full_deployment_scope
        if self.all_branch_deployments_scope is not None:
            result["allBranchDeploymentsScope"] = self.all_branch_deployments_scope
        if self.specific_branch_deployment_scope is not None:
            result["specificBranchDeploymentScope"] = self.specific_branch_deployment_scope
        if self.local_deployment_scope is not None:
            result["localDeploymentScope"] = self.local_deployment_scope
        return result
