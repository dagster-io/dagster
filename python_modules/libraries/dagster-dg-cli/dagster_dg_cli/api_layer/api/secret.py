"""Secret endpoints - REST-like interface."""

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional

from dagster_dg_cli.api_layer.graphql_adapter.secret import (
    get_secret_via_graphql,
    list_secrets_via_graphql,
)
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.secret import DgApiSecret, DgApiSecretList


@dataclass(frozen=True)
class DgApiSecretApi:
    """Secret API operations."""

    client: IGraphQLClient

    def list_secrets(
        self,
        location_name: Optional[str] = None,
        scope: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> "DgApiSecretList":
        """List secrets with optional filtering.

        Args:
            location_name: Optional filter by code location name
            scope: Optional scope filter ("deployment" or "organization")
            limit: Optional limit on number of results

        Returns:
            DgApiSecretList: List of secrets (values are never included for security)

        Note:
            Secret values are never exposed in list operations for security.
            Use get_secret with include_value=True to retrieve specific values.
        """
        return list_secrets_via_graphql(
            self.client,
            location_name=location_name,
            scope=scope,
            include_values=False,  # Never expose values in list operations
            limit=limit,
        )

    def get_secret(
        self,
        secret_name: str,
        location_name: Optional[str] = None,
        include_value: bool = False,
    ) -> "DgApiSecret":
        """Get a specific secret.

        Args:
            secret_name: Name of the secret to retrieve
            location_name: Optional filter by code location name
            include_value: Whether to include the secret value (default: False for security)

        Returns:
            DgApiSecret: Single secret with optional value

        Raises:
            ValueError: If secret not found
        """
        return get_secret_via_graphql(
            self.client,
            secret_name=secret_name,
            location_name=location_name,
            include_value=include_value,
        )
