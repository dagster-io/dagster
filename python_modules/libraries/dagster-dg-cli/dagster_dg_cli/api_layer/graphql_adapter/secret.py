"""GraphQL implementation for secret operations."""

from typing import TYPE_CHECKING, Any, Optional

from dagster_dg_cli.utils.plus.gql import (
    GET_SECRETS_FOR_SCOPES_QUERY,
    GET_SECRETS_FOR_SCOPES_QUERY_NO_VALUE,
)
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.secret import (
        DgApiSecret,
        DgApiSecretList,
        DgApiSecretScopesInput,
    )


def process_secrets_response(graphql_response: dict[str, Any]) -> "DgApiSecretList":
    """Process GraphQL response into DgApiSecretList.

    This is a pure function that can be easily tested without mocking GraphQL clients.

    Args:
        graphql_response: Raw GraphQL response containing secretsOrError

    Returns:
        DgApiSecretList: Processed secrets data

    Raises:
        ValueError: If GraphQL response contains errors
    """
    # Import pydantic models only when needed
    from dagster_dg_cli.api_layer.schemas.secret import DgApiSecret, DgApiSecretList

    secrets_or_error = graphql_response.get("secretsOrError")
    if not secrets_or_error:
        return DgApiSecretList(items=[], total=0)

    # Handle GraphQL error responses
    if secrets_or_error.get("__typename") == "UnauthorizedError":
        raise ValueError(f"Unauthorized: {secrets_or_error.get('message', 'Access denied')}")

    if secrets_or_error.get("__typename") == "PythonError":
        error_msg = secrets_or_error.get("message", "Unknown error")
        raise ValueError(f"Server error: {error_msg}")

    # Extract secrets array
    secrets_data = secrets_or_error.get("secrets", [])

    secrets = []
    for secret_data in secrets_data:
        # Convert GraphQL response to DgApiSecret model using field aliases
        secret = DgApiSecret.model_validate(secret_data)
        secrets.append(secret)

    return DgApiSecretList(
        items=secrets,
        total=len(secrets),
    )


def process_single_secret_response(
    graphql_response: dict[str, Any], secret_name: str
) -> "DgApiSecret":
    """Process GraphQL response for a single secret.

    Args:
        graphql_response: Raw GraphQL response
        secret_name: Name of the requested secret

    Returns:
        DgApiSecret: Single secret data

    Raises:
        ValueError: If secret not found or response contains errors
    """
    secret_list = process_secrets_response(graphql_response)

    if not secret_list.items:
        raise ValueError(f"Secret '{secret_name}' not found")

    if len(secret_list.items) > 1:
        # This shouldn't happen with secretName filter, but handle it
        matching_secrets = [s for s in secret_list.items if s.name == secret_name]
        if not matching_secrets:
            raise ValueError(f"Secret '{secret_name}' not found")
        return matching_secrets[0]

    return secret_list.items[0]


def build_secret_scopes_input(scope: Optional[str] = None) -> "DgApiSecretScopesInput":
    """Build SecretScopesInput based on scope parameter.

    Args:
        scope: Optional scope filter ("deployment" or "organization")

    Returns:
        DgApiSecretScopesInput: Scopes configuration for GraphQL query
    """
    from dagster_dg_cli.api_layer.schemas.secret import DgApiSecretScopesInput

    if scope == "deployment":
        # For deployment scope, include both full deployment and local deployment
        return DgApiSecretScopesInput(
            fullDeploymentScope=True,
            localDeploymentScope=True,
        )
    elif scope == "organization":
        # For organization scope, include all scopes
        return DgApiSecretScopesInput(
            fullDeploymentScope=True,
            allBranchDeploymentsScope=True,
            localDeploymentScope=True,
        )
    else:
        # Default: include all scopes
        return DgApiSecretScopesInput(
            fullDeploymentScope=True,
            allBranchDeploymentsScope=True,
            localDeploymentScope=True,
        )


def list_secrets_via_graphql(
    client: IGraphQLClient,
    location_name: Optional[str] = None,
    scope: Optional[str] = None,
    include_values: bool = False,
    limit: Optional[int] = None,
) -> "DgApiSecretList":
    """Fetch secrets using GraphQL.

    Args:
        client: GraphQL client
        location_name: Optional filter by code location
        scope: Optional scope filter ("deployment" or "organization")
        include_values: Whether to include secret values (security sensitive)
        limit: Optional limit on number of results

    Returns:
        DgApiSecretList: List of secrets

    Raises:
        ValueError: If GraphQL query fails or returns errors
    """
    scopes_input = build_secret_scopes_input(scope)

    # Choose query based on whether values should be included
    if include_values:
        query = GET_SECRETS_FOR_SCOPES_QUERY
    else:
        query = GET_SECRETS_FOR_SCOPES_QUERY_NO_VALUE

    variables: dict[str, Any] = {
        "scopes": scopes_input.to_dict(),
    }

    if location_name:
        variables["locationName"] = location_name

    result = client.execute(query, variables)

    secret_list = process_secrets_response(result)

    # Apply limit if specified
    if limit is not None:
        secret_list.items = secret_list.items[:limit]
        secret_list.total = len(secret_list.items)

    return secret_list


def get_secret_via_graphql(
    client: IGraphQLClient,
    secret_name: str,
    location_name: Optional[str] = None,
    include_value: bool = False,
) -> "DgApiSecret":
    """Fetch a specific secret using GraphQL.

    Args:
        client: GraphQL client
        secret_name: Name of the secret to fetch
        location_name: Optional filter by code location
        include_value: Whether to include secret value (security sensitive)

    Returns:
        DgApiSecret: Single secret

    Raises:
        ValueError: If secret not found or GraphQL query fails
    """
    # For get operations, we include all scopes to find the secret
    scopes_input = build_secret_scopes_input(scope=None)

    # Choose query based on whether value should be included
    if include_value:
        query = GET_SECRETS_FOR_SCOPES_QUERY
    else:
        query = GET_SECRETS_FOR_SCOPES_QUERY_NO_VALUE

    variables = {
        "scopes": scopes_input.to_dict(),
        "secretName": secret_name,
    }

    if location_name:
        variables["locationName"] = location_name

    result = client.execute(query, variables)

    return process_single_secret_response(result, secret_name)
