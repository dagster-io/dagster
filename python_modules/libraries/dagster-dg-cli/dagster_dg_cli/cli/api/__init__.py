"""API commands for interacting with Dagster Plus REST-like interface."""

from dagster_dg_cli.cli.api.cli_group import api_group

__all__ = ["api_group"]
