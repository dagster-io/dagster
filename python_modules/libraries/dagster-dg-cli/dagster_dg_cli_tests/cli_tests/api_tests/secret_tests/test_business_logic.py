"""Test secret business logic functions without mocks.

These tests focus on testing pure functions that process data without requiring
GraphQL client mocking or external dependencies.
"""

from datetime import datetime

from dagster_dg_cli.api_layer.schemas.secret import (
    DgApiSecret,
    DgApiSecretList,
    DgApiSecretScopesInput,
    DgApiUpdatedBy,
)
from dagster_dg_cli.cli.api.formatters import format_secret, format_secrets


class TestFormatSecrets:
    """Test the secret formatting functions."""

    def _create_sample_secret_list(self):
        """Create sample SecretList for testing."""
        secrets = [
            DgApiSecret(
                id="secret-1-uuid-12345",
                name="database_password",
                value="super_secret_value",  # Would be hidden in actual output
                location_names=["data_pipeline", "analytics"],
                full_deployment_scope=True,
                all_branch_deployments_scope=False,
                specific_branch_deployment_scope=None,
                local_deployment_scope=False,
                can_view_secret_value=True,
                can_edit_secret=True,
                updated_by=DgApiUpdatedBy(email="admin@company.com"),
                update_timestamp=datetime(
                    2022, 1, 1, 14, 20, 0
                ),  # UTC to avoid timezone edge cases
            ),
            DgApiSecret(
                id="secret-2-uuid-67890",
                name="api_key",
                value=None,  # No value - might not have permission to view
                location_names=[],  # All locations
                full_deployment_scope=False,
                all_branch_deployments_scope=True,
                specific_branch_deployment_scope=None,
                local_deployment_scope=False,
                can_view_secret_value=False,
                can_edit_secret=False,
                updated_by=None,
                update_timestamp=None,
            ),
            DgApiSecret(
                id="secret-3-uuid-abcdef",
                name="staging_token",
                value="staging_value_123",
                location_names=["staging_app"],
                full_deployment_scope=False,
                all_branch_deployments_scope=False,
                specific_branch_deployment_scope="staging",
                local_deployment_scope=True,
                can_view_secret_value=True,
                can_edit_secret=True,
                updated_by=DgApiUpdatedBy(email="dev@company.com"),
                update_timestamp=datetime(2022, 1, 1, 14, 21, 0),  # UTC
            ),
        ]
        return DgApiSecretList(items=secrets, total=3)

    def _create_empty_secret_list(self):
        """Create empty SecretList for testing."""
        return DgApiSecretList(items=[], total=0)

    def _create_single_secret(self):
        """Create single Secret for testing."""
        return DgApiSecret(
            id="single-secret-uuid-xyz",
            name="development_secret",
            value="dev_secret_value",
            location_names=["dev_pipeline", "test_pipeline"],
            full_deployment_scope=False,
            all_branch_deployments_scope=False,
            specific_branch_deployment_scope=None,
            local_deployment_scope=True,
            can_view_secret_value=True,
            can_edit_secret=True,
            updated_by=DgApiUpdatedBy(email="developer@company.com"),
            update_timestamp=datetime(2022, 1, 1, 14, 20, 0),
        )

    def test_format_secrets_text_output(self, snapshot):
        """Test formatting secrets as text."""
        from dagster_shared.utils.timing import fixed_timezone

        secret_list = self._create_sample_secret_list()
        with fixed_timezone("UTC"):
            result = format_secrets(secret_list, as_json=False)

        # Snapshot the entire text output
        snapshot.assert_match(result)

    def test_format_secrets_json_output(self, snapshot):
        """Test formatting secrets as JSON."""
        secret_list = self._create_sample_secret_list()
        result = format_secrets(secret_list, as_json=True)

        # For JSON, we want to snapshot the parsed structure to avoid formatting differences
        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_empty_secrets_text_output(self, snapshot):
        """Test formatting empty secret list as text."""
        secret_list = self._create_empty_secret_list()
        result = format_secrets(secret_list, as_json=False)

        snapshot.assert_match(result)

    def test_format_empty_secrets_json_output(self, snapshot):
        """Test formatting empty secret list as JSON."""
        secret_list = self._create_empty_secret_list()
        result = format_secrets(secret_list, as_json=True)

        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_single_secret_text_output(self, snapshot):
        """Test formatting single secret as text."""
        from dagster_shared.utils.timing import fixed_timezone

        secret = self._create_single_secret()
        with fixed_timezone("UTC"):
            result = format_secret(secret, as_json=False)

        snapshot.assert_match(result)

    def test_format_single_secret_text_output_with_value(self, snapshot):
        """Test formatting single secret as text with value shown."""
        from dagster_shared.utils.timing import fixed_timezone

        secret = self._create_single_secret()
        with fixed_timezone("UTC"):
            result = format_secret(secret, as_json=False, show_value=True)

        snapshot.assert_match(result)

    def test_format_single_secret_json_output(self, snapshot):
        """Test formatting single secret as JSON."""
        secret = self._create_single_secret()
        result = format_secret(secret, as_json=True)

        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_single_secret_json_output_with_value(self, snapshot):
        """Test formatting single secret as JSON with value shown."""
        secret = self._create_single_secret()
        result = format_secret(secret, as_json=True, show_value=True)

        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_secret_without_metadata(self, snapshot):
        """Test formatting secret with no metadata."""
        from dagster_shared.utils.timing import fixed_timezone

        secret = DgApiSecret(
            id="simple-secret-uuid",
            name="simple_secret",
            value="simple_value",
            location_names=[],
            full_deployment_scope=True,
            all_branch_deployments_scope=False,
            specific_branch_deployment_scope=None,
            local_deployment_scope=False,
            can_view_secret_value=True,
            can_edit_secret=True,
            updated_by=None,
            update_timestamp=None,
        )
        with fixed_timezone("UTC"):
            result = format_secret(secret, as_json=False)

        snapshot.assert_match(result)

    def test_format_secret_with_all_scopes(self, snapshot):
        """Test formatting secret with all scope types enabled."""
        from dagster_shared.utils.timing import fixed_timezone

        secret = DgApiSecret(
            id="multi-scope-secret-uuid",
            name="multi_scope_secret",
            value="multi_scope_value",
            location_names=["location1", "location2"],
            full_deployment_scope=True,
            all_branch_deployments_scope=True,
            specific_branch_deployment_scope="feature-branch",
            local_deployment_scope=True,
            can_view_secret_value=True,
            can_edit_secret=False,
            updated_by=DgApiUpdatedBy(email="system@company.com"),
            update_timestamp=datetime(2022, 1, 1, 14, 22, 0),
        )
        with fixed_timezone("UTC"):
            result = format_secret(secret, as_json=False)

        snapshot.assert_match(result)

    def test_format_secret_no_permissions(self, snapshot):
        """Test formatting secret with no permissions."""
        from dagster_shared.utils.timing import fixed_timezone

        secret = DgApiSecret(
            id="no-perm-secret-uuid",
            name="restricted_secret",
            value=None,  # No access to value
            location_names=["secure_location"],
            full_deployment_scope=False,
            all_branch_deployments_scope=False,
            specific_branch_deployment_scope=None,
            local_deployment_scope=False,
            can_view_secret_value=False,
            can_edit_secret=False,
            updated_by=DgApiUpdatedBy(email="admin@company.com"),
            update_timestamp=datetime(2022, 1, 1, 14, 23, 0),
        )
        with fixed_timezone("UTC"):
            result = format_secret(secret, as_json=False)

        snapshot.assert_match(result)


class TestSecretDataProcessing:
    """Test processing of secret data structures.

    This class tests any pure functions that process the raw GraphQL responses
    into our domain models and data model creation.
    """

    def test_secret_creation_with_all_scopes(self, snapshot):
        """Test creating secrets with all possible scope combinations."""
        secrets = []

        # Test different scope combinations
        scope_combinations = [
            {"full_deployment_scope": True, "all_branch_deployments_scope": False},
            {"full_deployment_scope": False, "all_branch_deployments_scope": True},
            {
                "full_deployment_scope": False,
                "all_branch_deployments_scope": False,
                "specific_branch_deployment_scope": "main",
            },
            {"local_deployment_scope": True},
            {
                "full_deployment_scope": True,
                "all_branch_deployments_scope": True,
                "local_deployment_scope": True,
            },
        ]

        for i, scopes in enumerate(scope_combinations):
            secret = DgApiSecret(
                id=f"scope-test-secret-{i}",
                name=f"scope_test_secret_{i}",
                value=f"test_value_{i}",
                location_names=[f"location_{i}"],
                full_deployment_scope=scopes.get("full_deployment_scope", False),
                all_branch_deployments_scope=scopes.get("all_branch_deployments_scope", False),
                specific_branch_deployment_scope=scopes.get(
                    "specific_branch_deployment_scope", None
                ),
                local_deployment_scope=scopes.get("local_deployment_scope", False),
                can_view_secret_value=True,
                can_edit_secret=True,
                updated_by=DgApiUpdatedBy(email=f"user{i}@company.com"),
                update_timestamp=datetime(2022, 1, 1, 14, 20, i),
            )
            secrets.append(secret)

        secret_list = DgApiSecretList(items=secrets, total=len(secrets))

        # Test JSON serialization works correctly for all scopes
        result = secret_list.model_dump_json(indent=2)
        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_secret_updated_by_handling(self):
        """Test secret updated_by entry creation and access."""
        secret = DgApiSecret(
            id="updated-by-test-secret",
            name="updated_by_test",
            value="test_value",
            location_names=["test_location"],
            full_deployment_scope=True,
            all_branch_deployments_scope=False,
            specific_branch_deployment_scope=None,
            local_deployment_scope=False,
            can_view_secret_value=True,
            can_edit_secret=True,
            updated_by=DgApiUpdatedBy(email="test@company.com"),
            update_timestamp=datetime(2022, 1, 1, 14, 20, 0),
        )

        assert secret.updated_by is not None
        assert secret.updated_by.email == "test@company.com"
        assert secret.update_timestamp == datetime(2022, 1, 1, 14, 20, 0)

    def test_secret_list_total_count(self):
        """Test that SecretList properly tracks total count."""
        secrets = [
            DgApiSecret(
                id=f"secret-{i}",
                name=f"secret_{i}",
                value=f"value_{i}",
                location_names=[],
                full_deployment_scope=True,
                all_branch_deployments_scope=False,
                specific_branch_deployment_scope=None,
                local_deployment_scope=False,
                can_view_secret_value=True,
                can_edit_secret=True,
                updated_by=None,
                update_timestamp=None,
            )
            for i in range(3)
        ]

        secret_list = DgApiSecretList(
            items=secrets, total=10
        )  # Total could be different from items length (pagination)

        assert len(secret_list.items) == 3
        assert secret_list.total == 10

    def test_secret_scopes_input_to_dict(self):
        """Test SecretScopesInput conversion to dict."""
        # Test with all fields
        scopes_input = DgApiSecretScopesInput(
            full_deployment_scope=True,
            all_branch_deployments_scope=False,
            specific_branch_deployment_scope="main",
            local_deployment_scope=True,
        )

        result = scopes_input.to_dict()
        expected = {
            "fullDeploymentScope": True,
            "allBranchDeploymentsScope": False,
            "specificBranchDeploymentScope": "main",
            "localDeploymentScope": True,
        }
        assert result == expected

        # Test with None values (should be excluded)
        scopes_input_sparse = DgApiSecretScopesInput(
            full_deployment_scope=True,
            all_branch_deployments_scope=None,
            specific_branch_deployment_scope=None,
            local_deployment_scope=False,
        )

        result_sparse = scopes_input_sparse.to_dict()
        expected_sparse = {
            "fullDeploymentScope": True,
            "localDeploymentScope": False,
        }
        assert result_sparse == expected_sparse

    def test_secret_permissions_combinations(self):
        """Test secret with various permission combinations."""
        permission_combos = [
            {"can_view_secret_value": True, "can_edit_secret": True},
            {"can_view_secret_value": True, "can_edit_secret": False},
            {"can_view_secret_value": False, "can_edit_secret": False},
            {"can_view_secret_value": False, "can_edit_secret": True},  # Unusual but possible
        ]

        for i, perms in enumerate(permission_combos):
            secret = DgApiSecret(
                id=f"perm-test-secret-{i}",
                name=f"permission_test_{i}",
                value="test_value" if perms["can_view_secret_value"] else None,
                location_names=[],
                full_deployment_scope=True,
                all_branch_deployments_scope=False,
                specific_branch_deployment_scope=None,
                local_deployment_scope=False,
                can_view_secret_value=perms["can_view_secret_value"],
                can_edit_secret=perms["can_edit_secret"],
                updated_by=None,
                update_timestamp=None,
            )

            assert secret.can_view_secret_value == perms["can_view_secret_value"]
            assert secret.can_edit_secret == perms["can_edit_secret"]

    def test_secret_value_hiding_in_json_format(self):
        """Test that secret values are properly hidden in JSON output when requested."""
        secret = DgApiSecret(
            id="value-hiding-test",
            name="secret_with_value",
            value="super_secret_password",
            location_names=["secure_app"],
            full_deployment_scope=True,
            all_branch_deployments_scope=False,
            specific_branch_deployment_scope=None,
            local_deployment_scope=False,
            can_view_secret_value=True,
            can_edit_secret=True,
            updated_by=None,
            update_timestamp=None,
        )

        # Test with show_value=False (default) - should hide value
        result_hidden = format_secret(secret, as_json=True, show_value=False)
        import json

        parsed_hidden = json.loads(result_hidden)
        assert parsed_hidden["value"] == "<hidden>"

        # Test with show_value=True - should show actual value
        result_shown = format_secret(secret, as_json=True, show_value=True)
        parsed_shown = json.loads(result_shown)
        assert parsed_shown["value"] == "super_secret_password"
