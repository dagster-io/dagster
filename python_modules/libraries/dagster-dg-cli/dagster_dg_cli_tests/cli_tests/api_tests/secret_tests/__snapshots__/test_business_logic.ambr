# serializer version: 1
# name: TestFormatSecrets.test_format_empty_secrets_json_output[0]
  dict({
    'items': list([
    ]),
    'total': 0,
  })
# ---
# name: TestFormatSecrets.test_format_empty_secrets_text_output[0]
  'No secrets found.'
# ---
# name: TestFormatSecrets.test_format_secret_no_permissions[0]
  '''
  Name: restricted_secret
  Locations: secure_location
  Scopes: None
  Permissions:
    Can Edit: No
    Can View Value: No
  
  Value: <hidden - use --show-value to display>
  
  Updated By: admin@company.com
  Updated: 2022-01-01 14:23:00
  '''
# ---
# name: TestFormatSecrets.test_format_secret_with_all_scopes[0]
  '''
  Name: multi_scope_secret
  Locations: location1, location2
  Scopes: Full Deployment, All Branch Deployments, Branch: feature-branch, Local Deployment
  Permissions:
    Can Edit: No
    Can View Value: Yes
  
  Value: <hidden - use --show-value to display>
  
  Updated By: system@company.com
  Updated: 2022-01-01 14:22:00
  '''
# ---
# name: TestFormatSecrets.test_format_secret_without_metadata[0]
  '''
  Name: simple_secret
  Locations: All code locations
  Scopes: Full Deployment
  Permissions:
    Can Edit: Yes
    Can View Value: Yes
  
  Value: <hidden - use --show-value to display>
  '''
# ---
# name: TestFormatSecrets.test_format_secrets_json_output[0]
  dict({
    'items': list([
      dict({
        'all_branch_deployments_scope': False,
        'can_edit_secret': True,
        'can_view_secret_value': True,
        'full_deployment_scope': True,
        'id': 'secret-1-uuid-12345',
        'local_deployment_scope': False,
        'location_names': list([
          'data_pipeline',
          'analytics',
        ]),
        'name': 'database_password',
        'specific_branch_deployment_scope': None,
        'update_timestamp': '2022-01-01T14:20:00',
        'updated_by': dict({
          'email': 'admin@company.com',
        }),
        'value': 'super_secret_value',
      }),
      dict({
        'all_branch_deployments_scope': True,
        'can_edit_secret': False,
        'can_view_secret_value': False,
        'full_deployment_scope': False,
        'id': 'secret-2-uuid-67890',
        'local_deployment_scope': False,
        'location_names': list([
        ]),
        'name': 'api_key',
        'specific_branch_deployment_scope': None,
        'update_timestamp': None,
        'updated_by': None,
        'value': None,
      }),
      dict({
        'all_branch_deployments_scope': False,
        'can_edit_secret': True,
        'can_view_secret_value': True,
        'full_deployment_scope': False,
        'id': 'secret-3-uuid-abcdef',
        'local_deployment_scope': True,
        'location_names': list([
          'staging_app',
        ]),
        'name': 'staging_token',
        'specific_branch_deployment_scope': 'staging',
        'update_timestamp': '2022-01-01T14:21:00',
        'updated_by': dict({
          'email': 'dev@company.com',
        }),
        'value': 'staging_value_123',
      }),
    ]),
    'total': 3,
  })
# ---
# name: TestFormatSecrets.test_format_secrets_text_output[0]
  '''
  Name: database_password
  Locations: data_pipeline, analytics
  Scopes: Full Deployment
  Can Edit: Yes
  Can View Value: Yes
  Updated By: admin@company.com
  Updated: 2022-01-01 14:20:00
  
  Name: api_key
  Locations: All code locations
  Scopes: All Branch Deployments
  Can Edit: No
  Can View Value: No
  
  Name: staging_token
  Locations: staging_app
  Scopes: Branch: staging, Local Deployment
  Can Edit: Yes
  Can View Value: Yes
  Updated By: dev@company.com
  Updated: 2022-01-01 14:21:00
  '''
# ---
# name: TestFormatSecrets.test_format_single_secret_json_output[0]
  dict({
    'all_branch_deployments_scope': False,
    'can_edit_secret': True,
    'can_view_secret_value': True,
    'full_deployment_scope': False,
    'id': 'single-secret-uuid-xyz',
    'local_deployment_scope': True,
    'location_names': list([
      'dev_pipeline',
      'test_pipeline',
    ]),
    'name': 'development_secret',
    'specific_branch_deployment_scope': None,
    'update_timestamp': '2022-01-01 14:20:00',
    'updated_by': dict({
      'email': 'developer@company.com',
    }),
    'value': '<hidden>',
  })
# ---
# name: TestFormatSecrets.test_format_single_secret_json_output_with_value[0]
  dict({
    'all_branch_deployments_scope': False,
    'can_edit_secret': True,
    'can_view_secret_value': True,
    'full_deployment_scope': False,
    'id': 'single-secret-uuid-xyz',
    'local_deployment_scope': True,
    'location_names': list([
      'dev_pipeline',
      'test_pipeline',
    ]),
    'name': 'development_secret',
    'specific_branch_deployment_scope': None,
    'update_timestamp': '2022-01-01T14:20:00',
    'updated_by': dict({
      'email': 'developer@company.com',
    }),
    'value': 'dev_secret_value',
  })
# ---
# name: TestFormatSecrets.test_format_single_secret_text_output[0]
  '''
  Name: development_secret
  Locations: dev_pipeline, test_pipeline
  Scopes: Local Deployment
  Permissions:
    Can Edit: Yes
    Can View Value: Yes
  
  Value: <hidden - use --show-value to display>
  
  Updated By: developer@company.com
  Updated: 2022-01-01 14:20:00
  '''
# ---
# name: TestFormatSecrets.test_format_single_secret_text_output_with_value[0]
  '''
  Name: development_secret
  Locations: dev_pipeline, test_pipeline
  Scopes: Local Deployment
  Permissions:
    Can Edit: Yes
    Can View Value: Yes
  
  Value:
    dev_secret_value
  
  Updated By: developer@company.com
  Updated: 2022-01-01 14:20:00
  '''
# ---
# name: TestSecretDataProcessing.test_secret_creation_with_all_scopes[0]
  dict({
    'items': list([
      dict({
        'all_branch_deployments_scope': False,
        'can_edit_secret': True,
        'can_view_secret_value': True,
        'full_deployment_scope': True,
        'id': 'scope-test-secret-0',
        'local_deployment_scope': False,
        'location_names': list([
          'location_0',
        ]),
        'name': 'scope_test_secret_0',
        'specific_branch_deployment_scope': None,
        'update_timestamp': '2022-01-01T14:20:00',
        'updated_by': dict({
          'email': 'user0@company.com',
        }),
        'value': 'test_value_0',
      }),
      dict({
        'all_branch_deployments_scope': True,
        'can_edit_secret': True,
        'can_view_secret_value': True,
        'full_deployment_scope': False,
        'id': 'scope-test-secret-1',
        'local_deployment_scope': False,
        'location_names': list([
          'location_1',
        ]),
        'name': 'scope_test_secret_1',
        'specific_branch_deployment_scope': None,
        'update_timestamp': '2022-01-01T14:20:01',
        'updated_by': dict({
          'email': 'user1@company.com',
        }),
        'value': 'test_value_1',
      }),
      dict({
        'all_branch_deployments_scope': False,
        'can_edit_secret': True,
        'can_view_secret_value': True,
        'full_deployment_scope': False,
        'id': 'scope-test-secret-2',
        'local_deployment_scope': False,
        'location_names': list([
          'location_2',
        ]),
        'name': 'scope_test_secret_2',
        'specific_branch_deployment_scope': 'main',
        'update_timestamp': '2022-01-01T14:20:02',
        'updated_by': dict({
          'email': 'user2@company.com',
        }),
        'value': 'test_value_2',
      }),
      dict({
        'all_branch_deployments_scope': False,
        'can_edit_secret': True,
        'can_view_secret_value': True,
        'full_deployment_scope': False,
        'id': 'scope-test-secret-3',
        'local_deployment_scope': True,
        'location_names': list([
          'location_3',
        ]),
        'name': 'scope_test_secret_3',
        'specific_branch_deployment_scope': None,
        'update_timestamp': '2022-01-01T14:20:03',
        'updated_by': dict({
          'email': 'user3@company.com',
        }),
        'value': 'test_value_3',
      }),
      dict({
        'all_branch_deployments_scope': True,
        'can_edit_secret': True,
        'can_view_secret_value': True,
        'full_deployment_scope': True,
        'id': 'scope-test-secret-4',
        'local_deployment_scope': True,
        'location_names': list([
          'location_4',
        ]),
        'name': 'scope_test_secret_4',
        'specific_branch_deployment_scope': None,
        'update_timestamp': '2022-01-01T14:20:04',
        'updated_by': dict({
          'email': 'user4@company.com',
        }),
        'value': 'test_value_4',
      }),
    ]),
    'total': 5,
  })
# ---
